tensorboard/data/server/run.rs

```diff
@@ -86,16 +97,34 @@ impl TimeSeries {
             rsv: StageReservoir::new(capacity),
         }
     }
-}
 
-/// A value staged in the reservoir.
-///
-/// This is kept as close as possible to the on-disk event representation, since every record in
-/// the stream is converted into this format.
-#[derive(Debug)]
-struct StageValue {
-    wall_time: WallTime,
-    payload: EventValue,
+    /// Writes all staged data for this time series into the commit.
+    fn commit(&mut self, tag: &Tag, run: &mut commit::RunData) {
+        use pb::DataClass;
+        match self.data_class {
+            DataClass::Scalar => self.commit_to(tag, &mut run.scalars, EventValue::into_scalar),
+            DataClass::Tensor => todo!("tensor time series not yet supported"),
+            DataClass::BlobSequence => todo!("blob sequence time series not yet supported"),
+            _ => (),
+        };
+    }
+
+    /// Helper for `commit`: writes staged data for this time series into storage for a statically
+    /// known data class.
+    fn commit_to<V, F: FnMut(EventValue) -> Result<V, commit::DataLoss>>(
+        &mut self,
+        tag: &Tag,
+        store: &mut commit::TagStore<V>,
+        mut enrich: F,
+    ) {
+        let commit_ts = store
+            .entry(tag.clone())
```

Comment:
> But then `Tag` is a dynamically sized type,
> and I had a bit of trouble defining it (I tried, but didn’t put too much
> effort into it since I’m not certain that we _need_ it right now).

Asked: <https://stackoverflow.com/questions/64977525/how-can-i-newtype-str-and-string-in-safe-rust>

Reply:
Thanks for all the detail!  FWIW, I doubt there is going to be a particularly measurable performance impact from these clones so keeping the typing simpler is fine by me :)  Mostly just trying to cultivate the right intuition.

And yeah, `ToOwned` looks like the trait what I was suggesting would actually require.

Reply:
I have written out a draft to create `Run`/`RunBuf` and `Tag`/`TagBuf`.
It works. The API is definitely nicer; the implementation is definitely
more complicated. It uses one line of `unsafe` with a pretty trivial
safety justification.

[Here is the patch.][1] No need for a review, and I’m not planning to
send it out any time soon. (Not yet convinced that it’s worth it.) But
in case you want to see some macros or some `unsafe`, there you go.

[1]: https://gist.github.com/wchargin/1fb815c7106164027e184f9d9cd2b3ae#file-newtypes-run-tag-slices-patch-L259